Added Controller for IAM Group

rahulchheda opened the pull request:
- Added basic CRUP Operations
    - Added utility func's
    - Added new CR and respective CRD for IAMGroup
    - Added new CR and respective CRD for IAMGroupPolicyAttachment
    - Added new CR and respective CRD for IAMUserGroupAttachment
    - Added Unit test cases
    - Fixed Variable naming in user unit tests
Signed-off-by: Rahul M Chheda <rchheda@infracloud.io>

<!--
Thank you for helping to improve Crossplane!

We strongly recommend you look through our contributor guide at https://git.io/fj2m9
if this is your first time opening a Crossplane pull request. You can find us in
https://slack.crossplane.io/messages/dev if you need any help contributing.
-->

### Description of your changes
<!--
Briefly describe what this pull request does. Be sure to direct your reviewers'
attention to anything that needs special consideration.

We love pull requests that resolve an open Crossplane issue. If yours does, you
can uncomment the below line to indicate which issue your PR fixes, for example
"Fixes #500":

Fixes #
-->

### Checklist
<!--
Please run through the below readiness checklist. The first two items are
relevant to every Crossplane pull request.
-->
I have:
- [ ] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Ensured this PR contains a neat, self documenting set of commits.
- [ ] Updated any relevant [documentation], [examples], or [release notes].
- [ ] Updated the dependencies in [`app.yaml`] to include any new role permissions.

[documentation]: https://github.com/crossplane/crossplane/tree/master/docs
[examples]: https://github.com/crossplane/crossplane/tree/master/cluster/examples
[release notes]: https://github.com/crossplane/crossplane/tree/master/PendingReleaseNotes.md
[`app.yaml`]: https://github.com/crossplane/provider-aws/blob/master/config/package/manifests/app.yaml

sahil-lakhwani committed:
move network resources to v1beta1

Signed-off-by: sahil-lakhwani <sahilakhwani@gmail.com>

sahil-lakhwani committed:
network: user external name as identifier

Signed-off-by: sahil-lakhwani <sahilakhwani@gmail.com>

sahil-lakhwani committed:
network: add tags to resources

Signed-off-by: sahil-lakhwani <sahilakhwani@gmail.com>

sahil-lakhwani committed:
network/internetgateway: remove nested if

Signed-off-by: sahil-lakhwani <sahilakhwani@gmail.com>

sahil-lakhwani committed:
network/internetgateway: check for AlreadyAttached

Signed-off-by: sahil-lakhwani <sahilakhwani@gmail.com>

sahil-lakhwani committed:
network: make vpcId optional

Signed-off-by: sahil-lakhwani <sahilakhwani@gmail.com>

sahil-lakhwani committed:
Apply new references paattern

Signed-off-by: sahil-lakhwani <sahilakhwani@gmail.com>

sahil-lakhwani committed:
network: ignore reference and selector types while comparing

Signed-off-by: sahil-lakhwani <sahilakhwani@gmail.com>

sahil-lakhwani committed:
Add SQS resource

Signed-off-by: sahil-lakhwani <sahilakhwani@gmail.com>

arush-sal commented on go.mod:
```diff
@@ -19,6 +19,7 @@ require (
 	github.com/smartystreets/assertions v0.0.0-20180820201707-7c9eb446e3cf // indirect
 	github.com/smartystreets/goconvey v0.0.0-20180222194500-ef6db91d284a // indirect
 	github.com/stretchr/testify v1.4.0
+	golang.org/x/tools/gopls v0.4.1 // indirect
```
why do we need this?

rahulchheda replied:
Will remove it in newer commits, changing PR to WIP

sahil-lakhwani commented on apis/identity/v1alpha1/iamgroup_types.go:
```diff
@@ -0,0 +1,80 @@
+
+	// A list of tags that you want to attach to the newly created user.
+	// +optional
+	//Tags []Tag `json:"tags,omitempty"`
```
If not needed, we can remove these.

sahil-lakhwani commented on pkg/clients/iam/iamgroup.go:
```diff
@@ -0,0 +1,83 @@
+	}
+
+	return cmp.Equal(desired, &observed, cmpopts.IgnoreInterfaces(struct{ resource.AttributeReferencer }{})), nil
+}
```
I think since there's just one field under `forProvider`, we don't really need a method like this to see the difference.
You can just compare it in the controller itself. Also, please see  if the only field, `Path` is indeed modifiable.

sahil-lakhwani commented on pkg/controller/identity/iamgroup/controller.go:
```diff
@@ -0,0 +1,197 @@
+
+	group := *observed.Group
+	current := cr.Spec.ForProvider.DeepCopy()
+	iam.LateInitializeGroup(&cr.Spec.ForProvider, &group)
```
There's just one field that can be updated, I think we don't nee a method for that.

sahil-lakhwani reviewed (changes requested):
@rahulchheda 

- You'll have to add the API schema in `apis/aws.go`
- Register the controller in `pkg/controller/aws.go`
- Can you add a working example YAML for this under `example/iam`

hasheddan committed:
Bump build submodule to use go 1.14.4

Bumps submodule to use go 1.14.4 to pick up a
workaround for linux kernel bug described in
this isssue https://github.com/golang/go/issues/37436

Signed-off-by: hasheddan <georgedanielmangum@gmail.com>

hasheddan committed:
Pass fsGroup in install.yaml

This re-enables support for IAM Roles for
Service Accounts by passing an fsGroup,
meaning that the projected service account
token volume will be owned by the GID and
the GID will be added as a supplemental group
to the container process.

Signed-off-by: hasheddan <georgedanielmangum@gmail.com>

rahulchheda commented:
go test results:

`ok      github.com/crossplane/provider-aws/pkg/controller/identity/iamgroup     0.027s  coverage: 88.7% of statements`
`ok      github.com/crossplane/provider-aws/pkg/clients/iam      0.012s  coverage: 28.0% of statements`

hasheddan committed:
Add TODO for removing fsGroup

Signed-off-by: hasheddan <georgedanielmangum@gmail.com>

jbw976 committed:
Merge pull request #240 from hasheddan/bumpbuildgo1.14.4

Bump build submodule to use go 1.14.4

jbw976 committed:
Merge pull request #242 from hasheddan/passfsgroup

Pass fsGroup in install.yaml

sahil-lakhwani commented on apis/identity/v1alpha1/iamgroup_types.go:
```diff
@@ -0,0 +1,75 @@
+
+// An IAMGroupSpec defines the desired state of an IAM Group.
+type IAMGroupSpec struct {
+	// +optional
+	runtimev1alpha1.ResourceSpec `json:",inline"`
+	// +optional
+	ForProvider IAMGroupParameters `json:"forProvider"`
```
I think these fields should not have `optional` annotation.

sahil-lakhwani commented on examples/aws-setup-provider/provider.yaml:
```diff
@@ -17,4 +17,4 @@ spec:
     name: aws-account-creds
     key: credentials
-  region: ((AWS_REGION))
+  region: us-east-1
```
Would you revert this?

sahil-lakhwani commented on pkg/clients/iam/iamgroup.go:
```diff
@@ -0,0 +1,75 @@
+func GenerateIAMGroup(in v1alpha1.IAMGroupParameters, role *iam.Group) error {
+
+	role.Path = in.Path
+	return nil
```
If you decide to remove `ISGroupUpToDate` which I think is unused, you can also remove this method.

rahulchheda replied:
Removed that func, but now I dont have any tests in `iamgroup_test.go`, can you suggest some?

sahil-lakhwani commented on pkg/clients/iam/iamgroup_test.go:
```diff
@@ -0,0 +1,55 @@
+	aws "github.com/crossplane/provider-aws/pkg/clients"
+)
+
+func TestIsGroupUpToDate(t *testing.T) {
```
Same here, if you remove `IsGroupUpToDate` this will not be needed.

sahil-lakhwani committed:
sqs: use tags map for comparision

Signed-off-by: sahil-lakhwani <sahilakhwani@gmail.com>

sahil-lakhwani committed:
sqs: rename controller file

Signed-off-by: sahil-lakhwani <sahilakhwani@gmail.com>

rahulchheda commented:
Resolved a few comments, can you please take a look again. Very concerned about `iamgroup_test.go`

sahil-lakhwani commented on pkg/clients/iam/iamusergroupattachment.go:
```diff
@@ -0,0 +1,35 @@
+	if group == nil {
+		return
+	}
+	in.GroupName = awsclients.LateInitializeString(in.GroupName, group.GroupName)
```
This method late-initialises just one field. Can we do that in the controller itself?

sahil-lakhwani commented on pkg/controller/identity/iamusergroupattachment/controller.go:
```diff
@@ -0,0 +1,200 @@
+func (e *external) Update(ctx context.Context, mgd resource.Managed) (managed.ExternalUpdate, error) {
+	// Updating any field will create a new User-Policy attachment in AWS, which will be
+	// irrelevant/out-of-sync to the original defined attachment.
+	// It is encouraged to instead create a new IAMUserPolicyAttachment resource.
```
Can you update the comment to be relevant to this PR?

sahil-lakhwani commented on pkg/controller/identity/iamusergroupattachment/controller.go:
```diff
@@ -0,0 +1,200 @@
+)
+
+const (
+	errNotUserInstance = "managed resource is not an User instance custom resource"
+
+	errCreateUserClient  = "cannot create IAM User client"
```
Can you change the above two strings to something relevant to this PR?

sahil-lakhwani reviewed (changes requested):
@rahulchheda 
- You'll have to register the new type in `apis/aws.go`
- In order to `runtimev1alpha1.Reference` types to resolve the value, there needs to a `ResolveReferences` method implemented by the enclosing struct. Refer: `apis/network/v1alpha3/referencers.go`
- In general, let's make sure that the code is in working condition at least for the simple cases.

sahil-lakhwani committed:
network: update CR status before loosing it

Signed-off-by: sahil-lakhwani <sahilakhwani@gmail.com>

sahil-lakhwani committed:
network: uppercase struct and field names

Signed-off-by: sahil-lakhwani <sahilakhwani@gmail.com>

sahil-lakhwani committed:
network/vpc: remove observation field

Signed-off-by: sahil-lakhwani <sahilakhwani@gmail.com>

ajaykangare1993 committed:
Added Certificate Authority code

Signed-off-by: ajaykangare <ajaykangare1993@gmail.com>

ajaykangare1993 committed:
Added Certificate Authority code

Signed-off-by: ajaykangare <ajaykangare1993@gmail.com>

sahil-lakhwani committed:
sqs: set external-name to queue name

Signed-off-by: sahil-lakhwani <sahilakhwani@gmail.com>

jbw976 committed:
governance: add OWNERS.md to define maintainers of this repo

Signed-off-by: Jared Watts <jbw976@gmail.com>

jbw976 committed:
Merge pull request #248 from jbw976/owners-update

governance: add OWNERS.md to define maintainers of this repo

hasheddan committed:
Bump aws-sdk-go-v2 to v0.23.0

Signed-off-by: hasheddan <georgedanielmangum@gmail.com>

hasheddan committed:
Add v1beta1 EKS Cluster API type

Signed-off-by: hasheddan <georgedanielmangum@gmail.com>

hasheddan committed:
Add EKS Cluster client

Signed-off-by: hasheddan <georgedanielmangum@gmail.com>

hasheddan committed:
Add v1beta1 EKS cluster controller

Signed-off-by: hasheddan <georgedanielmangum@gmail.com>

hasheddan committed:
Add EKS cluster example

Signed-off-by: hasheddan <georgedanielmangum@gmail.com>

hasheddan committed:
Add secret propagation controller for EKS v1beta1 Cluster

Signed-off-by: hasheddan <georgedanielmangum@gmail.com>

hasheddan committed:
Add target controller for EKS v1beta1 Cluster

Signed-off-by: hasheddan <georgedanielmangum@gmail.com>

hasheddan committed:
Check for nil EKS cluster when generating observation

Signed-off-by: hasheddan <georgedanielmangum@gmail.com>

ajaykangare1993 committed:
Resolved review comments and changed resource spec

Signed-off-by: ajaykangare <ajaykangare1993@gmail.com>

ajaykangare1993 committed:
Resolved review comments and changed resource spec

Signed-off-by: ajaykangare <ajaykangare1993@gmail.com>

sahil-lakhwani committed:
network/vpc: added vpcid printer column

Signed-off-by: sahil-lakhwani <sahilakhwani@gmail.com>

sahil-lakhwani committed:
sqs: set QUEUENAME printer column

Signed-off-by: sahil-lakhwani <sahilakhwani@gmail.com>

ajaykangare1993 committed:
Resolved tests errors

Signed-off-by: ajaykangare <ajaykangare1993@gmail.com>

ajaykangare1993 committed:
Resolved tests errors

Signed-off-by: ajaykangare <ajaykangare1993@gmail.com>

ajaykangare1993 committed:
Added controller for CA permission

Signed-off-by: ajaykangare <ajaykangare1993@gmail.com>

ajaykangare1993 committed:
Added controller for CA permission

Signed-off-by: ajaykangare <ajaykangare1993@gmail.com>

hasheddan committed:
Add notes and increase unit test coverage for v1beta1 EKS cluster

Signed-off-by: hasheddan <georgedanielmangum@gmail.com>

muvaf committed:
Merge pull request #182 from sahil-lakhwani/v1b1network

move network resources to v1beta1

muvaf committed:
Cosmetic changes on printer columns of network resources

Signed-off-by: Muvaffak Onus <onus.muvaffak@gmail.com>

muvaf committed:
Move network resources into ec2 group

Signed-off-by: Muvaffak Onus <onus.muvaffak@gmail.com>

hasheddan committed:
Update v1beta1 network unit tests

Fixes errors in v1beta1 network unit tests that result from bump of
aws-sdk-go-v2 bump to v0.23.0

Signed-off-by: hasheddan <georgedanielmangum@gmail.com>

muvaf committed:
Merge pull request #250 from muvaf/nw-move

Network resources in ec2 group

sahil-lakhwani committed:
sqs: change Value field in Tag to pointer

Signed-off-by: sahil-lakhwani <sahilakhwani@gmail.com>

hasheddan committed:
Late initialize and check tags in patch for EKS cluster update

Signed-off-by: hasheddan <georgedanielmangum@gmail.com>

hasheddan committed:
Change selectors to not be immutable and add EKS cluster version example

Signed-off-by: hasheddan <georgedanielmangum@gmail.com>

hasheddan committed:
Merge pull request #245 from hasheddan/eksv1beta1

EKS v1beta1 Cluster

ajaykangare1993 committed:
Changed directory as per group name acmpca

Signed-off-by: ajaykangare <ajaykangare1993@gmail.com>

ajaykangare1993 committed:
Changed directory as per group name acmpca

Signed-off-by: ajaykangare <ajaykangare1993@gmail.com>

ajaykangare1993 committed:
added missing dependancies

Signed-off-by: ajaykangare <ajaykangare1993@gmail.com>

ajaykangare1993 committed:
added missing dependancies

Signed-off-by: ajaykangare <ajaykangare1993@gmail.com>

hasheddan committed:
Remove extraneous group declarations

It appears during one of the kubebuilder transitions that these group
tags were added to the group-level package. In practice, we declare the
groups at the version level in each directory's doc.go. Therefore, these
tags do not impact generation and are not consistent across the repo.

Signed-off-by: hasheddan <georgedanielmangum@gmail.com>

hasheddan committed:
Merge pull request #252 from hasheddan/extraneous-group

Remove extraneous group declarations

hasheddan committed:
Fix panic in RouteTable client

The LateInitialize function for RouteTable is implemented incorrectly,
leading to the possibility of hitting array out-of-bounds panics. This
was likely missed because insufficient unit tests were provided for this
package. This PR is a simple fix to make us not panic, but in updating I
observed that there are fields missing from the RouteTable object and
it is also currently taking the place of multiple resource types in the
AWS API, such as Route and Association. These should likely be split out
into separate resources, but that fix is not provided in this PR. It
should also be noted that this PR does not fix the continual updates
that are made to a RouteTable despite the external resource being up to
date. If we do not break RouteTable into multiple resources then this
should be fixed.

Signed-off-by: hasheddan <georgedanielmangum@gmail.com>

hasheddan committed:
Make EKS cluster get ARN from IAM role reference

EKS Cluster was mistakenly resolving IAM role references by getting the
role name rather than its ARN. This updates the the reference resolver
and also make the roleArn omitempty so that updates that have a null ARN
(i.e. all of them before it is successfully resolved) will not fail. A
few instances of ID were also changed to Id.

Signed-off-by: hasheddan <georgedanielmangum@gmail.com>

hasheddan committed:
Move RouteTable to v1alpha4

This bumps the RouteTable type down to v1alpha4 as it does not have the
requirements of a beta level API.

Signed-off-by: hasheddan <georgedanielmangum@gmail.com>

hasheddan committed:
Merge pull request #255 from hasheddan/rt-bug

Fix panic in RouteTable client

hasheddan committed:
Merge pull request #256 from hasheddan/eks-cleanup

Make EKS cluster get ARN from IAM role reference

ajaykangare1993 committed:
Refactored certificate type

Signed-off-by: ajaykangare <ajaykangare1993@gmail.com>

ajaykangare1993 committed:
Refactored certificate type

Signed-off-by: ajaykangare <ajaykangare1993@gmail.com>

ajaykangare1993 committed:
Added todo comment

Signed-off-by: ajaykangare <ajaykangare1993@gmail.com>

ajaykangare1993 committed:
Added todo comment

Signed-off-by: ajaykangare <ajaykangare1993@gmail.com>

sahil-lakhwani committed:
sqs: add Retryer to fake client in controller tests

Signed-off-by: sahil-lakhwani <sahilakhwani@gmail.com>

ajaykangare1993 committed:
Rebase with master and resolved test error

Signed-off-by: ajaykangare <ajaykangare1993@gmail.com>

ajaykangare1993 committed:
Refactored package and type

Signed-off-by: ajaykangare <ajaykangare1993@gmail.com>

ajaykangare1993 committed:
Added changed acmpca crd

Signed-off-by: ajaykangare <ajaykangare1993@gmail.com>

ajaykangare1993 committed:
Updated build

Signed-off-by: ajaykangare <ajaykangare1993@gmail.com>

hasheddan committed:
Merge pull request #226 from ajaykangare/certificateauthority

ACMPCA Certificateauthority

muvaf committed:
Add IAMPolicy reference resolver to IAMRolePolicyAttachment resource

Signed-off-by: Muvaffak Onus <onus.muvaffak@gmail.com>

muvaf committed:
Add IAMPolicy reference resolver to IAMUserPolicyAttachment resource

Signed-off-by: Muvaffak Onus <onus.muvaffak@gmail.com>

muvaf committed:
IAMUserPolicyAttachment.spec.forProvider.userName should be immutable and non-pointer

Signed-off-by: Muvaffak Onus <onus.muvaffak@gmail.com>

muvaf committed:
Add missing IAM examples

Signed-off-by: Muvaffak Onus <onus.muvaffak@gmail.com>

muvaf committed:
Add categories to IAM managed resource CRDs

Signed-off-by: Muvaffak Onus <onus.muvaffak@gmail.com>

hasheddan committed:
Update ACMPCA types with omitempty in status

This adds omitempty json tags to the fields in the recently merged
ACMPCA types. It also fixes a number of typos and various logic across
the client functions.

Signed-off-by: hasheddan <georgedanielmangum@gmail.com>

hasheddan committed:
Update boilerplate header to 2020

This updates the boilerplate header to 2020 so that new generated
functions will have the appropriate date.

Signed-off-by: hasheddan <georgedanielmangum@gmail.com>

negz committed:
Merge pull request #222 from sahil-lakhwani/sqs

Add SQS resource

muvaf committed:
Merge pull request #257 from muvaf/pol-arn-reference

Policy ARN referencers for IAMUserPolicyAttachment and IAMRolePolicyAttachment

sahil-lakhwani committed:
network examples: change group name

Signed-off-by: sahil-lakhwani <sahilakhwani@gmail.com>

hasheddan committed:
Merge pull request #263 from sahil-lakhwani/ec2-examples

network examples: change group name

enderv committed:
Adding RDS Client tests

Signed-off-by: Chris <chris.vodak@gmail.com>

hasheddan committed:
Do not generate acmpca revocation configuration twice

Signed-off-by: hasheddan <georgedanielmangum@gmail.com>

hasheddan committed:
Uncomment cache referencers

A previous commit commented out the v1beta1 cache referencers.

Signed-off-by: hasheddan <georgedanielmangum@gmail.com>

hasheddan committed:
Merge pull request #258 from hasheddan/oeacmpca

Update ACMPCA types with omitempty in status

enderv committed:
Cleanup RDS Client tests-move shared vars to global

Signed-off-by: Chris <chris.vodak@gmail.com>

hasheddan committed:
Run generate on applicationintegration types

Generated headers were updated after the application integration types
were merged causing check-diff to fail on master branch.

Signed-off-by: hasheddan <georgedanielmangum@gmail.com>

hasheddan committed:
Merge pull request #266 from hasheddan/appint-gen

Run generate on applicationintegration types

hasheddan committed:
Merge pull request #265 from enderv/rds-testing

Adding RDS Client tests

Commit:
Added AWS IAMGroup with UserMembership, and PolicyAttachment
        - Added CRUD operations for various resources
        - Added test files for most of files (few exceptions are left, working on then actively)
        - Added examples for each resource
        - Added controllers for each resources

Signed-off-by: Rahul M Chheda <rchheda@infracloud.io>
